test/integration/idle_timeout_integration_test.cc

```diff
@@ -0,0 +1,161 @@
+  // this is the cause of flakes.
+  static constexpr uint64_t TimeoutMs = 200;
+};
+
+INSTANTIATE_TEST_CASE_P(Protocols, IdleTimeoutIntegrationTest,
+                        testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams()),
+                        HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+// Per-stream idle timeout when waiting for initial upstream headers.
+TEST_P(IdleTimeoutIntegrationTest, PerStreamIdleTimeoutBeforeUpstreamHeaders) {
+  auto response = setupPerStreamIdleTimeoutTest();
+
+  if (downstream_protocol_ == Http::CodecClient::Type::HTTP1) {
+    codec_client_->waitForDisconnect();
+  } else {
+    response->waitForReset();
+    codec_client_->close();
+  }
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_idle_timeout")->value());
+  EXPECT_TRUE(response->complete());
+  EXPECT_STREQ("408", response->headers().Status()->value().c_str());
+  EXPECT_EQ("stream timeout", response->body());
+}
+
+// Per-stream idle timeout when waiting for initial upstream headers after
+// sending data.
+TEST_P(IdleTimeoutIntegrationTest, PerStreamIdleTimeoutBeforeUpstreamHeadersAfterData) {
+  auto response = setupPerStreamIdleTimeoutTest();
+
+  Sleep();
+  codec_client_->sendData(*request_encoder_, 1, false);
+
+  if (downstream_protocol_ == Http::CodecClient::Type::HTTP1) {
+    codec_client_->waitForDisconnect();
+  } else {
+    response->waitForReset();
+    codec_client_->close();
+  }
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_EQ(1U, upstream_request_->bodyLength());
+
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_idle_timeout")->value());
+  EXPECT_TRUE(response->complete());
+  EXPECT_STREQ("408", response->headers().Status()->value().c_str());
+  EXPECT_EQ("stream timeout", response->body());
+}
+
+// Per-stream idle timeout after upstream headers have been sent.
+TEST_P(IdleTimeoutIntegrationTest, PerStreamIdleTimeoutAfterUpstreamHeaders) {
+  auto response = setupPerStreamIdleTimeoutTest();
+
+  upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+
+  if (downstream_protocol_ == Http::CodecClient::Type::HTTP1) {
+    codec_client_->waitForDisconnect();
+  } else {
+    response->waitForReset();
+    codec_client_->close();
+  }
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_idle_timeout")->value());
+  EXPECT_FALSE(response->complete());
+  EXPECT_STREQ("200", response->headers().Status()->value().c_str());
+  EXPECT_EQ("", response->body());
+}
+
+// Per-stream idle timeout after a sequence of header/data events.
+TEST_P(IdleTimeoutIntegrationTest, PerStreamIdleTimeoutAfterData) {
+  auto response = setupPerStreamIdleTimeoutTest();
+
+  Sleep();
```

Comment:
Mind giving this a flake check with
--jobs 60 --local_resources 100000000000,100000000000,10000000 --runs_per_test=1000
? I'd expect it to be flaky or slow, but I may just be biased against sleep tests :-P

Reply:
Yep, passes with:
```
INFO: Elapsed time: 282.940s, Critical Path: 19.66s
INFO: 1001 processes: 1001 linux-sandbox.
INFO: Build completed successfully, 1002 total actions
//test/integration:idle_timeout_integration_test                         PASSED in 19.5s
  Stats over 1000 runs: max = 19.5s, min = 14.7s, avg = 16.4s, dev = 0.9s
```